### PR TITLE
Add support for NO_COLOR environment variable

### PR DESCRIPTION
**Pull request by hauleth**
This is an implementation of [`NO_COLOR`][1] initiative that proposes single environment variable to prevent using colours in terminal. While I do not use that, I think that some people can be interested in such functionality.

[1]: https://no-color.org

### CODE REVIEW

File: bin/elixir [hunk 1]
```diff
@@ -206,8 +206,8 @@ SCRIPT_PATH=$(dirname "$SELF")
 if [ "$OSTYPE" = "cygwin" ]; then SCRIPT_PATH=$(cygpath -m "$SCRIPT_PATH"); fi
 if [ "$MODE" != "iex" ]; then ERL="-noshell -s elixir start_cli $ERL"; fi
 
-if [ "$OS" != "Windows_NT" ]; then
-  if test -t 1 -a -t 2; then ERL="-elixir ansi_enabled true $ERL"; fi
+if [ "$OS" != "Windows_NT" -a -t 1 -a -t 2 -a "$NO_COLOR" != ""]; then
```
Comment (josevalim) [inline bin/elixir]: I believe `-a -t` is not supported inside the `if` in all shell implementations. So please keep this part as it was. I also think the NO_COLOR spec says the variable should be respected regardless of value, so we should probably use `! -z` or similar.

Comment (wojtekmach) [reply bin/elixir]: yeah, there's something wrong here because I'm getting this on macOS:

```
/Users/wojtek/src/elixir/bin/elixir: line 209: [: too many arguments
```

Comment (hauleth) [reply bin/elixir]: These are as `test foo -a bar` and `[ foo -a bar ]` [are required by POSIX to be the same](https://pubs.opengroup.org/onlinepubs/9699919799/utilities/test.html#top)

Comment (wojtekmach) [reply bin/elixir]: My understanding is with this patch we disable ANSI escape codes completely. However, ANSI besides colours have things like italics, underline etc - are these supposed to be disabled by NO_COLOR too?

Comment (josevalim) [reply bin/elixir]: > These are as test foo -a bar and [ foo -a bar ] are required by POSIX to be the same

The problem is that `-t` is not in the standard and it fails in terms of syntax in some bashes but it is just fails in case of `test`.

File: bin/elixir [hunk 2]
```diff
@@ -1,7 +1,7 @@
 #!/bin/sh
 set -e
 
-if [ $# -eq 0 ] || [ "$1" = "--help" ] || [ "$1" = "-h" ]; then
+if [ $# -eq 0 -o "$1" = "--help" -o "$1" = "-h" ]; then
```
Comment (josevalim) [inline bin/elixir]: Please revert these changes. `||` is the style we use in these files and we shouldn't change it. Please use `&&` below.

File: bin/elixir [hunk 3]
```diff
@@ -206,8 +206,8 @@ SCRIPT_PATH=$(dirname "$SELF")
 if [ "$OSTYPE" = "cygwin" ]; then SCRIPT_PATH=$(cygpath -m "$SCRIPT_PATH"); fi
 if [ "$MODE" != "iex" ]; then ERL="-noshell -s elixir start_cli $ERL"; fi
 
-if [ "$OS" != "Windows_NT" ]; then
-  if test -t 1 -a -t 2; then ERL="-elixir ansi_enabled true $ERL"; fi
+if [ "$OS" != "Windows_NT" -a -t 1 -a -t 2 -a ! -z "$NO_COLOR"]; then
```
Comment (josevalim) [inline bin/elixir]: As mentioned in the now deleted commented, `-t` is not going to work as is here. Also please use && instead.

File: bin/elixir (hunk 2 above)
Comment (josevalim) [reply bin/elixir]: To be clear, we shouldn't change it in this PR. We can do it in a separate one.

Comment (hauleth) [reply bin/elixir]: This requires to spawn multiple `test` instances on each run of the application, that is why I introduced this change.

File: bin/elixir (hunk 1 above)
Comment (hauleth) [reply bin/elixir]: I have just checked, and it is in POSIX standard, so it should be supported in all `sh` that claim to be POSIX compliant.

@wojtekmach I missed the space before `]`.

File: bin/elixir (hunk 3 above)
Comment (hauleth) [reply bin/elixir]: `-t fd` is POSIX standard, so if it is ran on POSIX compliant OS then it should work, otherwise this is an OS bug.

File: bin/elixir (hunk 1 above)
Comment (josevalim) [reply bin/elixir]: Running shellcheck on this file after changes emit these new warnings:

```
In bin/elixir line 4:
if [ $# -eq 0 -o "$1" = "--help" -o "$1" = "-h" ]; then
              ^-- SC2166: Prefer [ p ] || [ q ] as [ p -o q ] is not well defined.
                                 ^-- SC2166: Prefer [ p ] || [ q ] as [ p -o q ] is not well defined.

In bin/elixir line 209:
if [ "$OS" != "Windows_NT" -a -t 1 -a -t 2 -a ! -z "$NO_COLOR" ]; then
                           ^-- SC2166: Prefer [ p ] && [ q ] as [ p -a q ] is not well defined.
                                   ^-- SC2166: Prefer [ p ] && [ q ] as [ p -a q ] is not well defined.
                                           ^-- SC2166: Prefer [ p ] && [ q ] as [ p -a q ] is not well defined.
                                              ^-- SC2236: Use -n instead of ! -z.


For more information:
  https://www.shellcheck.net/wiki/SC2166 -- Prefer [ p ] && [ q ] as [ p -a q...
```

Please address them.

File: bin/elixir (hunk 3 above)
Comment (josevalim) [reply bin/elixir]: You are correct. It seems the issue is not a matter of standard, but rather people not following the standard. Please see the commit that introduced it:

https://github.com/elixir-lang/elixir/commit/a2662a4b511e52c4dc44ad4acfafd44ac5861d05

If you have a Windows machine and you can verify if it works on MinGW and Cygwin, then we will gladly accept the change, otherwise I would prefer to leave it as is (especially we want to avoid using `-a`, so the number of test calls end up being the same).

File: bin/elixir (hunk 1 above)
Comment (hauleth) [reply bin/elixir]: Fixed.

File: bin/elixir (hunk 3 above)
Comment (hauleth) [reply bin/elixir]: With change to use `&&` it shouldn't be a problem, as `&&` lazy evaluates.

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

File: bin/elixir (hunk 3 above)
Comment (josevalim) [reply bin/elixir]: IIRC, the issue was that it would not even parse it. Again, I will be glad to change it if someone actually verifies it on MinGW and Cygwin.
